src/libraries/System.Private.CoreLib/src/System/IO/Strategies/UnixFileStreamStrategy.cs

```diff
@@ -0,0 +1,175 @@
+            base(path, mode, access, share, options, preallocationSize)
+        {
+        }
+
+        internal override bool IsAsync => _fileHandle.IsAsync;
+
+        public override IAsyncResult BeginRead(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
+            TaskToApm.Begin(ReadAsync(buffer, offset, count), callback, state);
+
+        public override int EndRead(IAsyncResult asyncResult) =>
+            TaskToApm.End<int>(asyncResult);
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+            ReadAsync(new Memory<byte>(buffer, offset, count), cancellationToken).AsTask();
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken)
+        {
+            if (!CanRead)
+            {
+                ThrowHelper.ThrowNotSupportedException_UnreadableStream();
+            }
+
+            if (CanSeek)
+            {
+                // This implementation updates the file position after the operation completes, rather than before.
+                // Also, unlike the Net5CompatFileStreamStrategy implementation, this implementation doesn't serialize operations.
+                ReadAsyncTaskSource rats = Interlocked.Exchange(ref _readAsyncTaskSource, null) ?? new ReadAsyncTaskSource(this);
+                return rats.QueueRead(destination, cancellationToken);
+            }
+
+            return RandomAccess.ReadAtOffsetAsync(_fileHandle, destination, fileOffset: -1, cancellationToken);
+        }
+
+        public override IAsyncResult BeginWrite(byte[] buffer, int offset, int count, AsyncCallback? callback, object? state) =>
+            TaskToApm.Begin(WriteAsync(buffer, offset, count), callback, state);
+
+        public override void EndWrite(IAsyncResult asyncResult) =>
+            TaskToApm.End(asyncResult);
+
+        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken) =>
+            WriteAsync(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();
+
+        public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
+        {
+            if (!CanWrite)
+            {
+                ThrowHelper.ThrowNotSupportedException_UnwritableStream();
+            }
+
+            long filePositionBefore = -1;
+            if (CanSeek)
+            {
+                filePositionBefore = _filePosition;
+                _filePosition += source.Length;
+            }
+
+#pragma warning disable CA2012 // The analyzer doesn't know the internal AsValueTask is safe.
+            return RandomAccess.WriteAtOffsetAsync(_fileHandle, source, filePositionBefore, cancellationToken).AsValueTask();
+#pragma warning restore CA2012
+        }
```

Comment:
the `Task` returning overload is going to go through: `.AsValueTask().AsTask()`. Can we reduce the number of casts?

```suggestion
        public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
            => WriteAsyncCore(new ReadOnlyMemory<byte>(buffer, offset, count), cancellationToken).AsTask();

        public override ValueTask WriteAsync(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
#pragma warning disable CA2012 // The analyzer doesn't know the internal AsValueTask is safe.
            => WriteAsyncCore(source, cancellationToken).AsValueTask();
#pragma warning restore CA2012

        private ValueTask<int> WriteAsyncCore(ReadOnlyMemory<byte> source, CancellationToken cancellationToken)
        {
            if (!CanWrite)
            {
                ThrowHelper.ThrowNotSupportedException_UnwritableStream();
            }

            long filePositionBefore = -1;
            if (CanSeek)
            {
                filePositionBefore = _filePosition;
                _filePosition += source.Length;
            }

            return RandomAccess.WriteAtOffsetAsync(_fileHandle, source, filePositionBefore, cancellationToken);
        }
```